catalog/resource_sql_table.go

```diff
@@ -294,29 +294,66 @@ func (ti *SqlTableInfo) buildTableCreateStatement() string {
 	return strings.Join(statements, "")
 }
 
-// Wrapping the column name with backtiks to avoid special character messing things up.
+// Wrapping the column name with backticks to avoid special character messing things up.
 func getWrappedColumnName(ci SqlColumnInfo) string {
 	return fmt.Sprintf("`%s`", ci.Name)
 }
 
 func (ti *SqlTableInfo) getStatementsForColumnDiffs(oldti *SqlTableInfo, statements []string, typestring string) []string {
-	// TODO: take out "force_new" in `column` and add case to handle addition and removal of columns.
-	for i, ci := range ti.ColumnInfos {
-		oldCi := oldti.ColumnInfos[i]
-		if ci.Name != oldCi.Name {
-			statements = append(statements, fmt.Sprintf("ALTER %s %s RENAME COLUMN %s to %s", typestring, ti.SQLFullName(), getWrappedColumnName(oldCi), getWrappedColumnName(ci)))
+	if len(ti.ColumnInfos) != len(oldti.ColumnInfos) {
+		nameToOldColumn := make(map[string]SqlColumnInfo)
+		nameToNewColumn := make(map[string]SqlColumnInfo)
+		for _, ci := range oldti.ColumnInfos {
+			nameToOldColumn[ci.Name] = ci
 		}
-		if ci.Comment != oldCi.Comment {
-			statements = append(statements, fmt.Sprintf("ALTER %s %s ALTER COLUMN %s COMMENT '%s'", typestring, ti.SQLFullName(), getWrappedColumnName(ci), parseComment(ci.Comment)))
+		for _, newCi := range ti.ColumnInfos {
+			nameToNewColumn[newCi.Name] = newCi
 		}
-		if ci.Nullable != oldCi.Nullable {
-			var keyWord string
-			if ci.Nullable {
-				keyWord = "DROP"
-			} else {
-				keyWord = "SET"
+
+		removeColumnStatements := make([]string, 0)
+
+		for name, oldCi := range nameToOldColumn {
+			if _, exists := nameToNewColumn[name]; !exists {
+				// Remove old column if old column is no longer found in the config.
+				removeColumnStatements = append(removeColumnStatements, getWrappedColumnName(oldCi))
+			}
+		}
+		if len(removeColumnStatements) > 0 {
+			removeColumnStatementsStr := strings.Join(removeColumnStatements, ", ")
+			statements = append(statements, fmt.Sprintf("ALTER %s %s DROP COLUMN IF EXISTS (%s)", typestring, ti.SQLFullName(), removeColumnStatementsStr))
+		}
+
+		addColumnStatements := make([]string, 0)
+
+		for name, newCi := range nameToNewColumn {
+			if _, exists := nameToOldColumn[name]; !exists {
+				// Add new column if new column is detected.
+				newCiStatement := ti.serializeColumnInfo(newCi)
+				addColumnStatements = append(addColumnStatements, newCiStatement)
+			}
+		}
+		if len(addColumnStatements) > 0 {
+			addColumnStatementsStr := strings.Join(addColumnStatements, ", ")
+			statements = append(statements, fmt.Sprintf("ALTER %s %s ADD COLUMN (%s)", typestring, ti.SQLFullName(), addColumnStatementsStr))
```

Comment:
Should we try to add it at the right position (i.e. `AFTER` the column that immediately precedes it)?

Reply:
Yeah good point, we should probably preserve the ordering